### PR TITLE
Allow arrows to take a Function to produce the vector

### PR DESCRIPTION
**Pull request by asinghvi17**
# Description

Allow `arrows` to take input of the form `x::AbstractVector, y::AbstractVector, [z::AbstractVector,] f::Function`

Add a description of your PR here.

## Type of change

Delete options that do not apply:

- [X] New feature (non-breaking change which adds functionality)

## Checklist

- [x] Added an entry in NEWS.md (for new features and breaking changes)
- [x] Added or changed relevant sections in the documentation
- [ ] Added unit tests for new algorithms, conversion methods, etc.
- [ ] Added reference image tests for new plotting functions, recipes, visual options, etc.

### CODE REVIEW

File: src/conversions.jl [hunk 1]
```diff
@@ -556,6 +556,25 @@ function convert_arguments(
...
+function convert_arguments(::Type{<: Arrows}, x::AbstractVector, y::AbstractVector, f::Function)
+    points = Point2f.(x, y')
+    f_out = f.(points)
+    return (vec(points), vec(Vec2f.(getindex.(f_out, 1), getindex.(f_out, 2))))
+end
+
+function convert_arguments(::Type{<: Arrows}, x::AbstractVector, y::AbstractVector, z::AbstractVector, f::Function)
+    points = [Point3f(x, y, z) for x in x, y in y, z in z]
+    f_out = f.(points)
+    return (vec(points), vec(Vec3f.(getindex.(f_out, 1), getindex.(f_out, 2), getindex.(f_out, 3)))
```
Comment (asinghvi17) [inline src/conversions.jl:576]:
```suggestion
    return (vec(points), vec(Vec3f.(getindex.(f_out, 1), getindex.(f_out, 2), getindex.(f_out, 3))))

File: docs/examples/plotting_functions/arrows.md [hunk 2]
```diff
@@ -93,3 +93,32 @@ arrows(
...
+Axis(f[1, 1], backgroundcolor = "black")
+
+xs = LinRange(0, 2pi, 20)
+ys = LinRange(0, 3pi, 20)
+# explicit method
+us = [sin(x) * cos(y) for x in xs, y in ys]
+vs = [-cos(x) * sin(y) for x in xs, y in ys]
+strength = vec(sqrt.(us .^ 2 .+ vs .^ 2))
+# function method
+f(x) = Point2f(sin(x[1])*cos(x[2]), -cos(x[1])*sin(x[2])
```
Comment (asinghvi17) [inline docs/examples/plotting_functions/arrows.md:116]:
```suggestion
f(x) = Point2f(sin(x[1])*cos(x[2]), -cos(x[1])*sin(x[2]))

File: docs/examples/plotting_functions/arrows.md [hunk 3]
```diff
@@ -93,3 +93,32 @@ arrows(
...
+us = [sin(x) * cos(y) for x in xs, y in ys]
+vs = [-cos(x) * sin(y) for x in xs, y in ys]
+strength = vec(sqrt.(us .^ 2 .+ vs .^ 2))
+# function method
+f(x) = Point2f(sin(x[1])*cos(x[2]), -cos(x[1])*sin(x[2]))
+
+arrows!(xs, ys, f, arrowsize = 10, lengthscale = 0.3,
+    arrowcolor = strength, linecolor = strength)
+
+f
```
Comment (asinghvi17) [inline docs/examples/plotting_functions/arrows.md:121]:
```suggestion
fig = Figure(resolution = (800, 800))
ax = Axis(fig[1, 1], backgroundcolor = "black")
xs = LinRange(0, 2pi, 20)
ys = LinRange(0, 3pi, 20)
# explicit method
us = [sin(x) * cos(y) for x in xs, y in ys]
vs = [-cos(x) * sin(y) for x in xs, y in ys]
strength = vec(sqrt.(us .^ 2 .+ vs .^ 2))
# function method
f(x) = Point2f(sin(x[1])*cos(x[2]), -cos(x[1])*sin(x[2]))
arrows!(ax, xs, ys, f, arrowsize = 10, lengthscale = 0.3,
    arrowcolor = strength, linecolor = strength)
fig
```

File: docs/examples/plotting_functions/arrows.md [hunk 4]
```diff
@@ -93,3 +93,29 @@ arrows(
...
+ax = Axis(fig[1, 1], backgroundcolor = "black")
+xs = LinRange(0, 2pi, 20)
+ys = LinRange(0, 3pi, 20)
+# explicit method
+us = [sin(x) * cos(y) for x in xs, y in ys]
+vs = [-cos(x) * sin(y) for x in xs, y in ys]
+strength = vec(sqrt.(us .^ 2 .+ vs .^ 2))
+# function method
+f(x) = Point2f(sin(x[1])*cos(x[2]), -cos(x[1])*sin(x[2]))
+arrows!(ax, xs, ys, f, arrowsize = 10, lengthscale = 0.3,
```
Comment (asinghvi17) [inline docs/examples/plotting_functions/arrows.md:116]:
```suggestion
arrow_fun(x) = Point2f(sin(x[1])*cos(x[2]), -cos(x[1])*sin(x[2]))
arrows!(ax, xs, ys, arrow_fun, arrowsize = 10, lengthscale = 0.3,

File: src/conversions.jl [hunk 5]
```diff
@@ -556,6 +556,25 @@ function convert_arguments(
...
+                        
+################################################################################
+#                                   <:Arrows                                   #
+################################################################################
+
+# Allow the user to pass a function to `arrows` which determines the direction
+# and magnitude of the arrows.  The function must accept `Point2f` as input.
+# and return Point2f or Vec2f or some array like structure as output.
+function convert_arguments(::Type{<: Arrows}, x::AbstractVector, y::AbstractVector, f::Function)
+    points = Point2f.(x, y')
```
Comment (ffreyer) [inline src/conversions.jl:568]: Shouldn't that be `Point2f.(x, y)`?

Comment (asinghvi17) [reply src/conversions.jl:568]: No, arrows always creates a matrix - the eventual positions are `Point2f.(x, y')` anyway. Plus the inputs are AbstractVectors so we know that we have to make the grid.

**Comment by ffreyer** [comment]
To me it seems more intuitive to just directly pass on args rather than doing a conversion beforehand. I.e.

```julia
arrows(xs, ys, (xs, ys) -> Vec2f.(-ys, xs))
```

rather than

```julia
arrows(xs, ys, ps -> Vec2f[(-p[2], p[1]) for p in ps])
```

That seems to be what surface and volume do.

I also wonder if we should just have a default conversion of `(args..., f) -> (args..., f(args...))` for all plot types.

**Comment by asinghvi17** [comment]
I do like the idea of that second conversion - but it might be easier to implement it per plot type.

**Pull request closed without merging**